### DISCUSSION TITLE
Project naming collisions

### DISCUSSION THREAD

**Discussion by khrome** [npm/feedback#357] [Registry] [answered]
**Handwavey Problem**
So I've had some naming fun where things intersected within a few characters of a valid name even though the repository shows it as available (showing "not available" instead of "not found" would go a long way to fixing this continual issue). 

**Specific Problem**
However today I had the most frustrating manifestation of this issue: If I maintain `some-library` and I attempt to create `some-library-cli` I get a vague error and the unhelpful suggestion to just add a namespace :P . 

**Potential Solution**
1) Publish the policy (not "similar names", but something specific like: the name can't match another repo, lowercased and stripped of specials AND must have a levenshtein distance of 3 from another module or ..."anything that stops the emails from MS corp" )
2) Make either the GUI show collisions (probably best, most obvious and simplest) or an endpoint where people can understand what is and isn't available before hitting `publish`

Can I get some specificity on what exactly the policy here is? Is there an endpoint we can use to understand "collisions" and policy with regards to this?

**Comment by khrome** [comment]
Found it. Someone published `somelibrarycli` 3 months ago. awesome :P 

This process does not need to be this painful.

**Comment by MylesBorins** [comment, marked as answer]
Hey @khrome. The primary reason for blocking these types of publishes is to avoid ["typosquating"](https://en.wikipedia.org/wiki/Typosquatting) attacks. There is a balance to be had here in transparency to ensure that bad-actors don't undermine our existing algorithms to make more sophisticated and potentially automated attacks. It is a constant cat-and-mouse game.

We are continuing to invest and improve the way in which we determine if we are going to block a publish based on typosquatting and as part of this effort I will work with the team to review and improve our error messages. Thank you for reporting

**Reply by khrome** [reply]
I remain skeptical about typosquatting packages as an actual attack vector (I'm pretty sure if a person bitshifted a popular repo they could fish for the org that pulled it (through organic memory errors), but I remain skeptical about human error exploits, at least until I see a package in the wild who's DL counts couldn't be 100% automated DLs from npm mirrors). 

I do understand the innate cat and mouse of security efforts, though, and the desire to avoid any liability as the delivery point of the attack. Thanks for putting in a effort around the potential error message modification!
